File: server/src/main/java/com/willkamp/vial/implementation/Assembly.java
package com.willkamp.vial.implementation;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.fasterxml.jackson.module.paramnames.ParameterNamesModule;
import com.willkamp.vial.api.VialServer;
import lombok.Getter;

public enum Assembly {
    instance;

    @Getter(lazy = true)
    private final VialConfig vialConfig = new VialConfig();

    @Getter(lazy = true)
    private final ObjectMapper objectMapper = objectMapper();

    SslContextFactory getSslContextFactory() {
        return new SslContextFactory(getVialConfig());
    }

    ChannelConfig getChannelConfig() {
        return new ChannelConfig();
    }


    VialChannelInitializer getVialChannelInitializer() {
        return new VialChannelInitializer(getSslContextFactory().createSslContext());
    }

    public VialServer getVialServer() {
        return new VialServerImpl(getVialConfig(), getChannelConfig(), getVialChannelInitializer());
    }

    private ObjectMapper objectMapper() {
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.registerModule(new ParameterNamesModule());
        objectMapper.registerModule(new JavaTimeModule());
        return objectMapper;
    }
}

File: server/src/main/java/com/willkamp/vial/implementation/Assembly.kt
package com.willkamp.vial.implementation

import com.fasterxml.jackson.databind.json.JsonMapper
import com.fasterxml.jackson.module.kotlin.jsonMapper
import com.fasterxml.jackson.module.kotlin.kotlinModule
import com.willkamp.vial.api.ServerInitializer
import com.willkamp.vial.api.VialConfig
import com.willkamp.vial.api.VialServer
import io.netty.channel.ChannelHandler

internal object Assembly {

    val objectMapper: JsonMapper by lazy {
        jsonMapper {
            addModule(kotlinModule())
        }
    }

    private fun createVialChannelInitializer(vialConfig: VialConfig, sslContextFactory: SslContextFactory, routeRegistry: RouteRegistry): VialChannelInitializer {
        val sslContext = if (vialConfig.useTls) sslContextFactory.createSslContext() else null
        return VialChannelInitializer(sslContext, vialConfig, routeRegistry)
    }

    fun createVialServer(): VialServer {
        val routeRegistry = RouteRegistry()
        val config = VialConfig()
        return VialServerImpl(
                config,
                ChannelConfig(),
                createVialChannelInitializer(config, SslContextFactory(), routeRegistry),
                routeRegistry
        )
    }

    fun createCustomInitializer(channelInitializer: ChannelHandler): ServerInitializer {
        return NettyInitializer(channelInitializer, ChannelConfig(), VialConfig())
    }
}
